zeus/monitor/carbon.py

```diff
@@ -348,44 +341,54 @@ def _polling_process(
     )  # {window_key -> {hour -> {gpu/cpu index -> energy}}}
     keys = set()
 
     # record energy measurement
-    def _update_energy_measurements(key: str, hour_key: str):
-        measurement = zeus_monitor.end_window(key)
+    def _update_energy_measurements(key: str, datetime: datetime):
+        measurement = zeus_monitor.end_window(key, sync_execution=False)
 
         for gpu_index, energy_measurement in measurement.gpu_energy.items():
-            energy_measurements[key][hour_key][f"gpu_{gpu_index}"] = energy_measurement
+            energy_measurements[key][datetime][f"gpu_{gpu_index}"] = energy_measurement
 
         if measurement.cpu_energy:
             for cpu_index, energy_measurement in measurement.cpu_energy.items():
-                energy_measurements[key][hour_key][
+                energy_measurements[key][datetime][
                     f"cpu_{cpu_index}"
                 ] = energy_measurement
 
+        if measurement.dram_energy:
+            for dram_index, energy_measurement in measurement.dram_energy.items():
+                energy_measurements[key][datetime][
+                    f"dram_{dram_index}"
+                ] = energy_measurement
+
     # update cumulative carbon emissions
     def _update_carbon_emissions(key: str):
         carbon_intensities = carbon_intensity_provider.get_recent_carbon_intensity()
+        print(carbon_intensities)
+
+        for dt, measurement_map in energy_measurements[key].items():
+            for index, energy in measurement_map.items():
+                hardware_type, num_index = index.split("_")
 
-        for hour_key, carbon_intensity in carbon_intensities.items():
-            for gpu_index in zeus_monitor.gpu_indices:
-                # divide by 3600 to convert joules -> Wh
-                gpu_carbon_emissions[key][gpu_index] += (
-                    energy_measurements[key][hour_key][f"gpu_{gpu_index}"]
-                    / 3600
-                    * carbon_intensity
-                )
-
-            for cpu_index in zeus_monitor.cpu_indices:
-                # divide by 3600 to convert joules -> Wh
-                cpu_carbon_emissions[key][cpu_index] += (
-                    energy_measurements[key][hour_key][f"cpu_{cpu_index}"]
-                    / 3600
-                    * carbon_intensity
-                )
+                if hardware_type == "gpu":
+                    gpu_carbon_emissions[key][int(num_index)] += (
+                        energy / 3.6e6 * carbon_intensities[dt]
+                    )
+                elif hardware_type == "cpu":
+                    cpu_carbon_emissions[key][int(num_index)] += (
+                        energy / 3.6e6 * carbon_intensities[dt]
+                    )
+                elif hardware_type == "dram":
+                    dram_carbon_emissions[key][int(num_index)] += (
+                        energy / 3.6e6 * carbon_intensities[dt]
+                    )
```

Comment:
So, depending on the meaning of the `datetime` object part of the `CarbonIntensityProvider` API contract, this may or may not be correct. Here, for a `datetime` key in `energy_measurements` (this `datetime` object is from `hour_floor`, so it's probably the energy consumed by the *previous* one hour) is accessed. This is matched with the carbon intensity value of the `carbon_intensities` corresponding to the same `datetime` object used to access `energy_measurements`, meaning that it the carbon intensity value accessed must be for the past hour as well. This makes sense, but it is not apparent from the API contract.

Assuming that I understood things correctly, for the purposes of this PR, I think documenting this assumption in the `CarbonIntensityProvider` method is sufficient. But, in an ideal implementation, what the `CarbonIntensityProvider` API should provide is a dictionary that maps *time ranges* to *carbon intensity values* and the polling process should handle that correspondingly.

Reply:
In my code, hour_floor  is used as the key and denotes the beginning of the windows of energy measurements. At the end of that hour, the `ZeusMonitor.end_window()` for each active window is called and the measurement is record with the hour_floor used at the beginning of the window. Example: 4PM is hour_floor and when `ZeusMonitor.begin_window()` are called for every active window. At 5PM, `ZeusMonitor.end_window()` is called for every active window, but each measurement collected is recorded into `energy_measurements` with the key 4PM. Hence, wouldn’t using hour_floor as the key mean each datetime key refers to the energy consumed within that hour and not the energy consumed with the previous hour? (i.e. if datetime key was 4PM in `energy_measurements`, energy consumed is measured in the time range from 4PM to 5PM and not from 3PM to 4PM?)

Reply:
Then, is using `hour_floor` to access carbon intensity correct? Is that how data is returned from ElectricityMaps?